File: acmOtherMaterials/Ulm Local Contest/2002/solution/ease.cc
// Problem   Polygon Programming with Ease
// Algorithm Geometry
// Runtime   O(n)
// Author    Walter Guttmann
// Date      30.05.2002

#include <fstream>

#ifdef VERIFY
#define in cin
#else
ifstream in ("ease.in");
#endif

int main ()
{
  int n;
  while (in >> n)
  {
    double x[n], y[n];
    for (int i=0 ; i<n ; i++)
      in >> x[i] >> y[i];
    double xx = x[0], yy = y[0];
    for (int i=0 ; i<n ; i++)
      xx = x[i] + (x[i] - xx), yy = y[i] + (y[i] - yy);
    xx = (x[0] + xx) / 2.0, yy = (y[0] + yy) / 2.0;
    cout << n;
    for (int i=0 ; i<n ; i++)
      cout.form (" %f %f", xx, yy),
      xx = x[i] + (x[i] - xx), yy = y[i] + (y[i] - yy);
    cout << endl;
  }
  return 0;
}

File: acmOtherMaterials/Ulm Local Contest/2002/solution/ease.hs
-- Problem   Polygon Programming with Ease
-- Algorithm Geometry
-- Runtime   O(n)
-- Author    Walter Guttmann
-- Date      30.05.2002

import Complex

main :: IO ()
main = readFile "ease.in" >>=
       mapM_ (putStrLn . pretty . solve . parse) . cases . map read . words

type Case  = [Point]
type Point = Complex Double

cases :: [Double] -> [[Double]]
cases [] = []
cases (d:xs) = let (ys,zs) = splitAt (2 * round d) xs in ys : cases zs

parse :: [Double] -> Case
parse [] = []
parse (x:y:xs) = x:+y : parse xs

solve :: Case -> Case
solve ps = init qs
  where rotate x y = y + (y - x)
        p1 = head ps
        p2 = foldl1 rotate ps
        p3 = (p1 + p2) / 2.0
        qs = scanl rotate p3 ps

pretty :: Case -> String
pretty ps = show (length ps) ++
            concat [ " " ++ show x ++ " " ++ show y | x:+y <- ps ]

